pkg/reconciler/taskrun/validate_resources.go

```diff
@@ -311,19 +318,27 @@ func validateTaskRunResults(tr *v1beta1.TaskRun, resolvedTaskSpec *v1beta1.TaskS
 }
 
 // mismatchedTypesResults checks and returns all the mismatched types of emitted results against specified results.
-func mismatchedTypesResults(tr *v1beta1.TaskRun, specResults []v1beta1.TaskResult) map[string][]string {
-	neededTypes := make(map[string][]string)
-	providedTypes := make(map[string][]string)
-	// collect needed keys for object results
+func mismatchedTypesResults(tr *v1beta1.TaskRun, specResults []v1beta1.TaskResult) map[string]string {
+	neededTypes := make(map[string]string)
+	mismatchedTypes := make(map[string]string)
+	var filteredResults []v1beta1.TaskRunResult
+	// collect needed types for results
 	for _, r := range specResults {
-		neededTypes[r.Name] = append(neededTypes[r.Name], string(r.Type))
+		neededTypes[r.Name] = string(r.Type)
 	}
 
-	// collect provided keys for object results
+	// collect mismatched types for results, and correct results in filteredResults
 	for _, trr := range tr.Status.TaskRunResults {
-		providedTypes[trr.Name] = append(providedTypes[trr.Name], string(trr.Type))
+		needed, ok := neededTypes[trr.Name]
+		if ok && needed != string(trr.Type) {
+			mismatchedTypes[trr.Name] = fmt.Sprintf("task result is expected to be \"%v\" type but was initialized to a different type \"%v\"", needed, trr.Type)
+		} else {
+			filteredResults = append(filteredResults, trr)
```

Comment:
it looks like this else can happen if either the result was provided in the correct type, or the result was provided but not declared? is it OK to have results that are provided but not declared? Might be worth adding a comment here

Reply:
yes great point! This case should be captured in results validation at admission webhook.
But after testing it looks like we don't have this validation. I will open an issue to track this.

Reply:
the issue is here, https://github.com/tektoncd/pipeline/issues/6097
I'm collecting feedback in slack channel. If we agree with this we may fix it in another pr

Reply:
> is it OK to have results that are provided but not declared? 

We have not implemented a strong validation around producing results. At the same time, if the result is not declared but consumed would return an error. Thanks @Yongxuanzhang for creating a separate issue for it. We could certainly introduce some kind of validation on using `$(results.result-does-not-exist.path)` when `result-does-not-exist` is not declared as a task result.

Reply:
Thanks Yongxuan! would you mind adding a comment with a TODO linking to the issue you created?

Reply:
@Yongxuanzhang this should be ready to merge after you add a comment with TODO as @lbernick requested. @lbernick appreciate such detailed and careful review 🙏  🤗

Reply:
Thanks! Just added